Backend/Backend/Services/ReportService.cs

```diff
@@ -0,0 +1,148 @@
+        // Check if entity is valid 
+        if (!AcceptableReportTargets.Contains(report.TargetEntityName)) {
+            throw new InvalidOperationException("Invalid report target");
+        }
+        
+        // Check that reported by is valid 
+        if (await _db.Users.FindAsync(report.ReportedBy) is null) {
+            throw new InvalidOperationException("Invalid reporter");
+        }
+        
+        // Check that target entity exists
+        var target = await GetReportTargetFromId(report.TargetEntityName, report.TargetId);
+        if (target is null) {
+            throw new InvalidOperationException("Invalid target entity");
+        }
+        
+        // Check if report already exists
+        if (await _db.Reports.AnyAsync(r => r.TargetEntityName == report.TargetEntityName && r.TargetId == report.TargetId && r.ReportedBy == report.ReportedBy)) {
+            throw new InvalidOperationException("Report already exists");
+        }
+        await _db.Reports.AddAsync(report);
+        _db.Reports.Add(report);
+        await _db.SaveChangesAsync();
+
+        return target;
+    }
+    
+    public Task<ImmutableArray<Report>> GetPendingReports() {
+        return Task.Run(() => _db.Reports.ToImmutableArray()) ;
+    }
+    
+    public async Task ResolveReport(User admin, Guid reportId) {
+        Report? report = await _db.Reports.FindAsync(reportId);
+        if (report is null) {
+            throw new InvalidOperationException("Invalid report ID");
+        }
+        await DeleteReportedEntity(admin.Id, report.TargetEntityName, report.TargetId);
+        report.Status = Models.Report.ReportStatus.Resolved;
+        report.DeletedBy = admin.Id;       
+        _db.Reports.Remove(report); // Soft deleted
+        await _db.SaveChangesAsync();
+    }
+    
+    public async Task RejectReport(User admin, Guid reportId) {
+        Report? report = await _db.Reports.FindAsync(reportId);
+        if (report is null) {
+            throw new InvalidOperationException("Invalid report ID");
+        }
+        report.Status = Models.Report.ReportStatus.Rejected;
+        report.DeletedBy = admin.Id;       
+        _db.Reports.Remove(report); // Soft deleted report
+        await _db.SaveChangesAsync();
+    }
+    
+    private async Task<ISoftDeletable?> GetReportTargetFromId(string entityName, Guid id) {
+        return entityName switch {
+            nameof(User) => await _db.Users.FindAsync(id),
+            nameof(Podcast) => await _db.Podcasts.FindAsync(id),
+            nameof(Episode) => await _db.Episodes.FindAsync(id),
+            nameof(Comment) => await _db.Comments.FindAsync(id),
+            nameof(CommentReply) => await _db.CommentReplies.FindAsync(id),
+            _ => null
+        };
+    }
+    
+    private async Task DeleteReportedEntity(Guid deletedBy, string entityName, Guid id) {
+        switch (entityName) {
+            case nameof(User):
```

Comment:
If you are feeling a bit degenerate, you can replace this with raw sql query where the table name gets changed depending on the entity, not at all needed though

Reply:
Maybe something to consider in the future if we add more entity types

Reply:
True

Reply:
Getting it using SQL raw query is real hassle.  It requires using reflection and even then it doesn't work properly.